packages/arb-avm-cpp/data_storage/src/value/code.cpp

```diff
@@ -45,84 +50,152 @@ std::array<unsigned char, segment_key_size> segment_key(uint64_t segment_id) {
+        it = std::copy(big_ref_count_ptr,
+                       big_ref_count_ptr + sizeof(big_ref_count), it);
+
+        auto big_op_count = boost::endian::native_to_big(op_count);
+        auto big_op_count_ptr = reinterpret_cast<const char*>(&big_op_count);
+        it = std::copy(big_op_count_ptr,
+                       big_op_count_ptr + sizeof(big_op_count), it);
+
+        auto big_hash_count = boost::endian::native_to_big(hash_count);
+        auto big_hash_count_ptr =
+            reinterpret_cast<const char*>(&big_hash_count);
+        it = std::copy(big_hash_count_ptr,
+                       big_hash_count_ptr + sizeof(big_hash_count), it);
+
+        auto big_chunk_count = boost::endian::native_to_big(chunk_count);
+        auto big_chunk_count_ptr =
+            reinterpret_cast<const char*>(&big_chunk_count);
+        it = std::copy(big_chunk_count_ptr,
+                       big_chunk_count_ptr + sizeof(big_chunk_count), it);
+        return value;
+    }
+};
+
+struct RawOperation {
     OpCode opcode;
     std::optional<ParsedSerializedVal> parsed_immediate;
-    uint256_t next_hash;
 };
 
-RawCodePoint extractRawCodePoint(
-    std::vector<unsigned char>::const_iterator& it) {
-    bool is_immediate = *it;
-    ++it;
-    auto opcode = static_cast<OpCode>(*it);
-    ++it;
-    auto ptr = reinterpret_cast<const char*>(&*it);
-    uint256_t next_hash = deserializeUint256t(ptr);
-    it += 32;
+RawOperation extractRawOperation(const char*& buf) {
+    bool is_immediate = *buf;
+    ++buf;
+    auto opcode = static_cast<OpCode>(*buf);
+    ++buf;
     if (!is_immediate) {
-        return {opcode, std::nullopt, next_hash};
+        return {opcode, std::nullopt};
     }
-    return {opcode, parseRecord(it), next_hash};
+    return {opcode, parseRecord(buf)};
 }
 
-std::vector<RawCodePoint> extractRawCodeSegment(
-    const std::vector<unsigned char>& stored_value) {
-    auto iter = stored_value.cbegin();
-    auto ptr = reinterpret_cast<const char*>(&*iter);
-    auto cp_count = deserialize_uint64_t(ptr);
-    iter += sizeof(cp_count);
-    std::vector<RawCodePoint> cps;
-    cps.reserve(cp_count);
-    for (uint64_t i = 0; i < cp_count; i++) {
-        cps.push_back(extractRawCodePoint(iter));
+void extractRawOperations(std::vector<RawOperation>& cps, const char*& buf) {
+    auto op_count = deserialize_uint64_t(buf);
+    for (uint64_t i = 0; i < op_count; i++) {
+        cps.push_back(extractRawOperation(buf));
     }
-    return cps;
 }
 
-void serializeCodePoint(const CodePoint& cp,
-                        std::vector<unsigned char>& serialized_code) {
-    // Ignore referemces to other code segments
-    serialized_code.push_back(cp.op.immediate ? 1 : 0);
-    serialized_code.push_back(static_cast<unsigned char>(cp.op.opcode));
-    marshal_uint256_t(cp.nextHash, serialized_code);
+void extractRawHashes(std::vector<uint256_t>& hashes, const char* ptr) {
```

Comment:
IMO it's a bit dangerous that this doesn't take in a reference to the pointer whereas `extractRawOperations` does

Reply:
Agreed. Modified `extractRawHashes` to take the pointer by reference. At some point it might be good to move away from using in-out params for this, but that would be a much bigger lift